one/src/events.rs

```diff
@@ -48,17 +71,214 @@ async fn slurp(opts: SlurpOpts) -> Result<()> {
+    let signatures_0 = envelope
+        .get_key("signatures")
+        .unwrap()
+        .get_index(0)
+        .unwrap();
+    let protected = signatures_0
+        .get_key("protected")
+        .unwrap()
+        .as_bytes()
+        .unwrap()
+        .as_slice();
+    // Deserialize protected as JSON
+    let protected_json: serde_json::Value = serde_json::from_slice(protected)?;
+    let controller = protected_json.get("kid").unwrap().as_str().unwrap();
+    let protected: Base64UrlString = protected.into();
+    let signature: Base64UrlString = signatures_0
+        .get_key("signature")
+        .unwrap()
+        .as_bytes()
+        .unwrap()
+        .as_slice()
+        .into();
+    let payload: Base64UrlString = envelope
+        .get_key("payload")
+        .unwrap()
+        .as_bytes()
+        .unwrap()
+        .as_slice()
+        .into();
+    let compact = format!("{}.{}.{}", protected, payload, signature);
+    let did = DidDocument::new(controller);
+    let jwk = Jwk::new(&did).await.unwrap();
+    ssi::jws::decode_verify(&compact, &jwk)?;
+    validate_data_event_payload(cid, block_store, Some(did.id.clone())).await?;
+    Ok(did.id)
+}
+
+// TODO: Validate the Data Event payload structure
+// TODO: Validate CACAO
+async fn validate_data_event_payload(
+    _payload_cid: &Cid,
+    _block_store: &SQLiteBlockStore,
+    _controller: Option<String>,
+) -> Result<()> {
+    Ok(())
+}
+
+// To validate a Time Event, we need to prove:
+// - TimeEvent/prev == TimeEvent/proof/root/${TimeEvent/path}
+// - TimeEvent/proof/root is in the Root Store
+//
+// - If root not in local root store try to read it from txHash.
+// - Validated time is the time from the Root Store
+async fn validate_time_event(
+    cid: &Cid,
+    block_store: &SQLiteBlockStore,
+    root_store: &SQLiteRootStore,
+    eth_rpc: &impl EthRpc,
+) -> Result<i64> {
+    if let Some(block) = block_store.get(cid.to_owned()).await? {
+        let time_event = CborValue::parse(&block)?;
+        // Destructure the proof to get the tag and the value
+        let proof_cid: Cid = time_event.path(&["proof"]).try_into()?;
+        let prev: Cid = time_event.path(&["prev"]).try_into()?;
+        if let Some(proof_block) = block_store.get(proof_cid).await? {
+            let proof_cbor = CborValue::parse(&proof_block)?;
+            let proof_root: Cid = proof_cbor.path(&["root"]).try_into()?;
+            let path: String = time_event.path(&["path"]).try_into()?;
+
+            // If prev not in root then TimeEvent is not valid.
+            if !prev_in_root(prev, proof_root, path, block_store).await? {
+                return Err(anyhow!("prev {} not in root {}", prev, proof_root));
+            }
+
+            // if root in root_store return timestamp.
+            // note: at some point we will need a negative cache to exponentially backoff eth_getTransactionByHash
+            if let Ok(Some(timestamp)) = root_store.get(proof_root.hash().digest()).await {
```

Comment:
Here we're only comparing the root CID to what we validated previously and put into the root store, but it's possible that the last time we saw this root it came in a TimeEvent with a different txHash.  If we see a new TimeEvent where the txHash doesn't match the root given, but the root is one we've seen before, we'd consider that valid when it should be considered invalid.

I think the root_store needs to also store the txHash associated with that root, and we need to double-check it matches here before we assume we can re-use the data in the root store without validating directly against ethereum.

Reply:
This makes sense the role of the `txHash` is to be a key into the root store. Whether that root store is SQLite or Ethereum Should not matter.

```sql
CREATE TABLE IF NOT EXISTS roots (
  txHash BLOB, 
  root BLOB, 
  timestamp INTEGER, 
  PRIMARY KEY(txHash)
);
```